File: packages/provide-uterm-platform/src/provide/uterm/pty/capture.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
from __future__ import annotations

import asyncio
import os
import struct
from dataclasses import dataclass
from pathlib import Path

from provide.telemetry import get_logger

from provide.uterm.pty.socket_utils import validate_socket_path

logger = get_logger(__name__)

CHANNEL_STDOUT = 0x01
CHANNEL_STDIN = 0x02
CHANNEL_CONNECT = 0x03

_HEADER = struct.Struct(">BI")  # channel (1B) + length (4B big-endian)

# Upper bound on buffered capture frames. A fast/local client on the capture
# socket would otherwise grow the queue without bound and OOM the host; once
# full we drop the oldest frame (the reader is the only consumer and capture is
# best-effort, so dropping beats blocking the connection handler).
_QUEUE_MAXSIZE = 4096

# Upper bound on a single frame's payload. The 4-byte length field admits up to
# ~4 GiB, and ``readexactly(length)`` accumulates that many bytes before
# enqueueing — so a malicious or buggy producer on the (owner-only) socket could
# announce a huge length and drive the host into OOM. Legitimate frames are at
# most a single intercepted ``write()`` (capture.c emits one frame per write,
# kilobytes in practice), so we cap well above any real frame and treat an
# over-cap length as a framing violation: drop the connection (PLAT-cap).
_MAX_FRAME_BYTES = 16 * 1024 * 1024  # 16 MiB

# Owner-only permissions for the listening Unix socket.
_SOCKET_MODE = 0o600
# umask that yields 0o600 at file creation (0o777 & ~0o177 == 0o600). Set around
# the bind so the socket is owner-only the instant it appears — see start().
_BIND_UMASK = 0o177


@dataclass
class CaptureFrame:
    channel: int
    data: bytes


class CaptureSocket:
    """
    Async Unix domain socket server that receives frames from libuterm_capture.

    Frame wire format: [1B channel][4B length big-endian][N bytes payload]
    """

    def __init__(self, socket_path: str) -> None:
        validate_socket_path(socket_path)
        self._path = socket_path
        self._server: asyncio.Server | None = None
        self._queue: asyncio.Queue[CaptureFrame] = asyncio.Queue(maxsize=_QUEUE_MAXSIZE)

    async def start(self) -> None:
        # Restrict the listening socket to the owner so other local users can't
        # connect and inject/observe captured terminal traffic. Set the umask
        # *before* the bind so the socket is created 0o600 atomically: a
        # post-bind chmod would leave a window where the socket exists with
        # default-umask perms (e.g. srwxr-xr-x) that any local user could
        # connect to. Restore the previous umask in finally so it always
        # happens.
        # NOTE: os.umask is process-global and not async-safe — keep the
        # set→bind→restore window as tight as possible (only the bind call).
        prev_umask = os.umask(_BIND_UMASK)
        try:
            self._server = await asyncio.start_unix_server(self._handle_connection, path=self._path)
        finally:
            os.umask(prev_umask)
        # Belt-and-suspenders: enforce 0o600 even if the platform ignored the
        # umask for AF_UNIX sockets. With the umask in place this is a no-op.
        os.chmod(self._path, _SOCKET_MODE)  # noqa: PTH101 — chmod the just-bound socket fd path

    async def stop(self) -> None:
        if self._server is None:
            return
        self._server.close()
        await self._server.wait_closed()
        self._server = None
        try:
            Path(self._path).unlink()
        except FileNotFoundError:
            pass

    async def read_frame(self) -> CaptureFrame:
        return await self._queue.get()

    def read_nowait(self) -> CaptureFrame | None:
        """Return the next buffered frame without blocking, or None when empty."""
        try:
            return self._queue.get_nowait()
        except asyncio.QueueEmpty:
            return None

    def _enqueue(self, frame: CaptureFrame) -> None:
        """Buffer *frame*, applying drop-oldest backpressure when the queue is
        full so a fast producer cannot grow it without bound (PLAT-cap)."""
        try:
            self._queue.put_nowait(frame)
        except asyncio.QueueFull:
            try:
                self._queue.get_nowait()
            except asyncio.QueueEmpty:  # pragma: no cover — full then empty is racy/unreachable single-threaded
                pass
            self._queue.put_nowait(frame)
            logger.warning("capture_backpressure_drop_oldest", maxsize=self._queue.maxsize)

    async def _handle_connection(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
        try:
            while True:
                header_bytes = await reader.readexactly(_HEADER.size)
                channel, length = _HEADER.unpack(header_bytes)
                if length > _MAX_FRAME_BYTES:
                    # Hostile/corrupt producer: refuse to allocate gigabytes for
                    # one frame. Drop the connection rather than read the body —
                    # checking BEFORE readexactly is what prevents the OOM.
                    logger.warning("capture_frame_too_large", length=length, cap=_MAX_FRAME_BYTES)
                    break
                data = await reader.readexactly(length)
                self._enqueue(CaptureFrame(channel=channel, data=data))
        except (asyncio.IncompleteReadError, ConnectionResetError):
            pass
        finally:
            writer.close()
            try:
                await writer.wait_closed()
            except Exception:  # noqa: S110
                pass

File: packages/provide-uterm-go/pty/capture.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package pty

import (
	"context"
	"encoding/binary"
	"errors"
	"io"
	"net"
	"os"
	"sync"

	ptel "github.com/provide-io/provide-telemetry/go"
)

// Capture frame channel identifiers. Port of capture.py CHANNEL_* constants.
const (
	ChannelStdout  = 0x01
	ChannelStdin   = 0x02
	ChannelConnect = 0x03
)

const (
	// headerSize is the framing header width: 1B channel + 4B big-endian length.
	headerSize = 5
	// queueMaxSize bounds buffered capture frames (drop-oldest past this). Port
	// of capture.py _QUEUE_MAXSIZE.
	queueMaxSize = 4096
	// maxFrameBytes caps a single frame's payload (framing-violation past this).
	// Port of capture.py _MAX_FRAME_BYTES.
	maxFrameBytes = 16 * 1024 * 1024 // 16 MiB
	// socketMode is the owner-only permission for the listening socket.
	socketMode = 0o600
	// bindUmask yields 0o600 at file creation (0o777 &^ 0o177 == 0o600).
	bindUmask = 0o177
)

// CaptureFrame is one decoded capture frame. Port of capture.CaptureFrame.
type CaptureFrame struct {
	Channel int
	Data    []byte
}

// CaptureSocket is a Unix-domain socket server that receives length-prefixed
// frames from libuterm_capture. Port of capture.CaptureSocket.
//
// Wire format: [1B channel][4B length big-endian][N bytes payload].
type CaptureSocket struct {
	path     string
	queueCap int

	mu       sync.Mutex
	listener net.Listener
	conns    map[net.Conn]struct{}
	started  bool
	closed   bool
	wg       sync.WaitGroup

	// queue is a buffered channel; enqMu makes drop-oldest atomic against
	// concurrent producers so overflow deterministically discards the oldest.
	queue chan CaptureFrame
	enqMu sync.Mutex
}

// NewCaptureSocket builds a CaptureSocket bound (on Start) to socketPath.
func NewCaptureSocket(socketPath string) (*CaptureSocket, error) {
	return newCaptureSocketWithQueue(socketPath, queueMaxSize)
}

// newCaptureSocketWithQueue is the queue-size-parameterized constructor used by
// tests to force overflow deterministically.
func newCaptureSocketWithQueue(socketPath string, queueCap int) (*CaptureSocket, error) {
	if err := ValidateSocketPath(socketPath); err != nil {
		return nil, err
	}
	return &CaptureSocket{
		path:     socketPath,
		queueCap: queueCap,
		conns:    make(map[net.Conn]struct{}),
		queue:    make(chan CaptureFrame, queueCap),
	}, nil
}

// Path returns the socket path.
func (s *CaptureSocket) Path() string { return s.path }

// Start binds the listening socket (owner-only) and begins accepting. Port of
// CaptureSocket.start.
func (s *CaptureSocket) Start() error {
	// Set a restrictive umask *before* bind so the socket is created 0o600
	// atomically — a post-bind chmod would leave a window where the socket
	// exists with default-umask perms that any local user could connect to.
	prev := umaskSet(bindUmask)
	ln, err := net.Listen("unix", s.path)
	umaskSet(prev)
	if err != nil {
		return err
	}
	// Belt-and-suspenders: enforce 0o600 even if the platform ignored the umask.
	if err := os.Chmod(s.path, socketMode); err != nil {
		_ = ln.Close()
		return err
	}
	s.mu.Lock()
	s.listener = ln
	s.started = true
	s.mu.Unlock()

	s.wg.Add(1)
	go s.acceptLoop(ln)
	return nil
}

// acceptLoop accepts connections until the listener is closed.
func (s *CaptureSocket) acceptLoop(ln net.Listener) {
	defer s.wg.Done()
	for {
		conn, err := ln.Accept()
		if err != nil {
			return
		}
		s.mu.Lock()
		if s.closed {
			s.mu.Unlock()
			_ = conn.Close()
			return
		}
		s.conns[conn] = struct{}{}
		s.mu.Unlock()

		s.wg.Add(1)
		go func() {
			defer s.wg.Done()
			defer s.removeConn(conn)
			s.handleConn(conn)
		}()
	}
}

func (s *CaptureSocket) removeConn(conn net.Conn) {
	s.mu.Lock()
	delete(s.conns, conn)
	s.mu.Unlock()
	_ = conn.Close()
}

// handleConn reads framed data from r until EOF or a framing violation. It is
// separated from the net.Conn so the framing logic is unit-testable with any
// io.Reader. Port of CaptureSocket._handle_connection's read loop.
func (s *CaptureSocket) handleConn(r io.Reader) {
	header := make([]byte, headerSize)
	for {
		if _, err := io.ReadFull(r, header); err != nil {
			return
		}
		channel := int(header[0])
		length := binary.BigEndian.Uint32(header[1:])
		if length > maxFrameBytes {
			// Hostile/corrupt producer: refuse to allocate for one frame. Drop
			// the connection rather than read the body — checking BEFORE the
			// body read is what prevents the OOM.
			ptel.GetLogger(context.Background(), "provide.uterm.pty").Warn(
				"capture_frame_too_large", "length", length, "cap", maxFrameBytes)
			return
		}
		data := make([]byte, length)
		if _, err := io.ReadFull(r, data); err != nil {
			return
		}
		s.enqueue(CaptureFrame{Channel: channel, Data: data})
	}
}

// enqueue buffers frame, applying drop-oldest backpressure when the queue is
// full. Port of CaptureSocket._enqueue.
func (s *CaptureSocket) enqueue(frame CaptureFrame) {
	s.enqMu.Lock()
	defer s.enqMu.Unlock()
	select {
	case s.queue <- frame:
		return
	default:
	}
	// Full: drop the oldest, then enqueue.
	select {
	case <-s.queue:
	default:
	}
	s.queue <- frame
	ptel.GetLogger(context.Background(), "provide.uterm.pty").Warn(
		"capture_backpressure_drop_oldest", "maxsize", s.queueCap)
}

// ReadFrame blocks until a frame is available or ctx is done. Port of
// CaptureSocket.read_frame.
func (s *CaptureSocket) ReadFrame(ctx context.Context) (CaptureFrame, error) {
	select {
	case f := <-s.queue:
		return f, nil
	case <-ctx.Done():
		return CaptureFrame{}, ctx.Err()
	}
}

// ReadNowait returns the next buffered frame and true, or false when empty.
// Port of CaptureSocket.read_nowait.
func (s *CaptureSocket) ReadNowait() (CaptureFrame, bool) {
	select {
	case f := <-s.queue:
		return f, true
	default:
		return CaptureFrame{}, false
	}
}

// QueueLen returns the number of buffered frames (test/observability helper).
func (s *CaptureSocket) QueueLen() int { return len(s.queue) }

// Stop closes the listener + open connections and removes the socket file.
// Idempotent. Port of CaptureSocket.stop.
func (s *CaptureSocket) Stop() error {
	s.mu.Lock()
	if !s.started || s.closed {
		// stop() before start() — or a second stop() — is a no-op and must not
		// remove any file (matches CaptureSocket.stop's `if server is None`).
		s.mu.Unlock()
		return nil
	}
	s.closed = true
	ln := s.listener
	s.listener = nil
	conns := make([]net.Conn, 0, len(s.conns))
	for c := range s.conns {
		conns = append(conns, c)
	}
	s.mu.Unlock()

	if ln != nil {
		_ = ln.Close()
	}
	for _, c := range conns {
		_ = c.Close()
	}
	s.wg.Wait()

	if err := os.Remove(s.path); err != nil && !errors.Is(err, os.ErrNotExist) {
		return err
	}
	return nil
}
